x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/create_rule_exceptions_route.ts

```diff
@@ -0,0 +1,275 @@
+              listsClient,
+            });
+          } else {
+            // This means that there was missed cleanup when this rule exception list was
+            // deleted and it remained referenced on the rule. Let's remove it from the rule,
+            // and update the rule's exceptions lists to include newly created default list.
+            const defaultList = await createAndAssociateDefaultExceptionList({
+              rule,
+              rulesClient,
+              listsClient,
+              removeOldAssociation: true,
+            });
+
+            createdItems = await createExceptionListItems({ items, defaultList, listsClient });
+          }
+        } else {
+          const defaultList = await createAndAssociateDefaultExceptionList({
+            rule,
+            rulesClient,
+            listsClient,
+            removeOldAssociation: false,
+          });
+
+          createdItems = await createExceptionListItems({ items, defaultList, listsClient });
+        }
+
+        const [validated, errors] = validate(createdItems, t.array(exceptionListItemSchema));
+        if (errors != null) {
+          return siemResponse.error({ body: errors, statusCode: 500 });
+        } else {
+          return response.ok({ body: validated ?? {} });
+        }
+      } catch (err) {
+        const error = transformError(err);
+        return siemResponse.error({
+          body: error.message,
+          statusCode: error.statusCode,
+        });
+      }
+    }
+  );
+};
+
+export const createExceptionListItems = async ({
+  items,
+  defaultList,
+  listsClient,
+}: {
+  items: CreateRuleExceptionListItemSchemaDecoded[];
+  defaultList: ExceptionListSchema;
+  listsClient: ExceptionListClient | null;
+}) => {
+  return Promise.all(
+    items.map((item) =>
+      listsClient?.createExceptionListItem({
+        comments: item.comments,
+        description: item.description,
+        entries: item.entries,
+        itemId: item.item_id,
+        listId: defaultList.list_id,
+        meta: item.meta,
+        name: item.name,
+        namespaceType: defaultList.namespace_type,
+        osTypes: item.os_types,
+        tags: item.tags,
+        type: item.type,
+      })
+    )
+  );
+};
+
+export const createAndAssociateDefaultExceptionList = async ({
+  rule,
+  listsClient,
+  rulesClient,
+  removeOldAssociation,
+}: {
+  rule: SanitizedRule<RuleParams>;
+  listsClient: ExceptionListClient | null;
+  rulesClient: RulesClient;
+  removeOldAssociation: boolean;
+}): Promise<ExceptionListSchema> => {
+  const exceptionList: CreateExceptionListSchema = {
+    description: `Exception list containing exceptions for rule with id: ${rule.id}`,
+    meta: undefined,
+    name: `Exceptions for rule - ${rule.name}`,
+    namespace_type: 'single',
+    tags: ['default_rule_exception_list'],
+    type: ExceptionListTypeEnum.RULE_DEFAULT,
+    version: 1,
+  };
+
+  // The `as` defeated me. Please send help
```

Comment:
Let's pair on this and figure it out; I have defeated the io-ts typings previously 💪 .

Reply:
Often these casts become necessary when the io-ts schema applies default values as part of the validation process - when that happens, the static type that io-ts generates for the schema is not as strict as the actual type at runtime, e.g. a field that gets an empty array by default might be `string[] | undefined` in the io-ts schema, even though at runtime we know due to the default value after validation it will be `string[]`. The `Decoded` types represent this, but require explicitly casting the object after validation and maintaining the separate `Decoded` types in sync with any default values.

I know we've had default values being applied in the schemas for a long time, but I've been trying a different approach of moving the default values outside the io-ts schemas, and it seems to work well IMO. It reduces the complexity of the schemas in return for explicitly applying the default values in the routes. For the rule schemas, this happens in `rule_converters.ts` and the defaults are enforced with static types. I don't know if that's the best strategy here, but it could help.

If there were no default values applied during `createExceptionListSchema.decode` then the type of the object that comes out of that `decode` would be `t.TypeOf<typeof createExceptionListSchema>` and no cast would be needed, but you'd have to apply the default values explicitly after decoding.

Reply:
Yea, that's exactly what's happening here. The `list_id` is being set by the decode. I looked at the pattern you mentioned and I think that does make sense. I don't feel super strongly about it here (if I add a comment explaining the cast) but can update or at least leave a note about coming back to it.

Reply:
I don't think the `as` is worth blocking this effort over; a small comment explaining is better than nothing, and we can look into creating an issue for the defaulting/typing issue if it becomes enough of a problem.